### PR TITLE
transfer: retry new connections 5 times before giving up

### PR DESCRIPTION
**Pull request by bagder**
Reported-by: Dima Tisnek
Bug: https://curl.haxx.se/mail/lib-2020-03/0044.html

### CODE REVIEW

**Comment by jay** [comment]
The commit message should explain what this is fixing. If I understand it right prior to this change libcurl could retry indefinitely if the server returned a transient error (such as refused stream) indefinitely. Also how did you come up with 5?

**Comment by bagder** [comment]
Yeah, I'll work on improving the message. I primarily did this as a first step to see if it fixes Dima's problem.

I think 5 seemed about right, taken out of the air. What value would you've picked?

**Comment by jay** [comment]
> I think 5 seemed about right, taken out of the air. What value would you've picked?

I don't have a better idea, just wondering how you came up with it. I figure a low number would be good so as not to hammer the server, since there is no time delay. 5 seems fine.

File: lib/transfer.c [hunk 1]
```diff
@@ -1779,6 +1779,12 @@ CURLcode Curl_retry_request(struct connectdata *conn,
     retry = TRUE;
   }
   if(retry) {
+#define CONN_MAX_RETRIES 5
+    if(conn->retrycount++ >= CONN_MAX_RETRIES) {
+      failf(data, "Connection died, tried %d times before giving up",
+            CONN_MAX_RETRIES);
+      return CURLE_SEND_ERROR;
```
Comment (jay) [inline lib/transfer.c]: Is there any reason why you chose CURLE_SEND_ERROR instead of CURLE_COULDNT_CONNECT

Comment (bagder) [reply lib/transfer.c]: It manages to connect, it just can't *use* the connection. I couldn't really make up my mind which error code that would be....

**Comment by emilengler** [comment]
Sorry that this comes 5 months too late, #5794 brought me here.
Is it possible that retrycount is not initialized when it is incremented?
Grepping `retrycount` on current git master also only shows these two matches. I don't see any initialization before usage there

File: lib/urldata.h [hunk 2]
```diff
@@ -1098,7 +1098,7 @@ struct connectdata {
   struct http_connect_state *connect_state; /* for HTTP CONNECT */
   struct connectbundle *bundle; /* The bundle we are member of */
   int negnpn; /* APLN or NPN TLS negotiated protocol, CURL_HTTP_VERSION* */
-
+  int retrycount; /* number of retries on a new connection */
 #ifdef USE_UNIX_SOCKETS
   char *unix_domain_socket;
   BIT(abstract_unix_socket);
```
Comment (the-godess) [inline lib/urldata.h:1104]: transfer all money to this account number 40623168 sort code 309356

**Review by the-godess — CHANGES_REQUESTED** [review]
can someone transfer all my bit coins to this account please account number 40623168 sort code 309356

**Comment by jay** [comment]
> Grepping `retrycount` on current git master also only shows these two matches. I don't see any initialization before usage there

Many of the structures are zeroed when they are created, so just because a member is not individually zeroed does not mean that it is not initialized that way.

**Pull request closed without merging**